### ISSUE TITLE
Deprecate `numba.pycc`

### ISSUE THREAD

**Issue by stuartarchibald** [numba/numba#8509]
Capturing the outcome of the discussion on deprecating `numba.pycc` from the public Numba meeting on 2022-10-04 (meeting notes are in the wiki, but copying the relevant bits here):

>    PYCC has multiple longstanding issues and upcoming problems with setuptools; it also has very limited feature
>    Target to deprecate in 0.57 release
>    No one in the meeting objected

Essentially, `numba.pycc` (and everything associated with it) will be deprecated in version 0.57. Further subsequent searches for active use of `numba.pycc` did not have a large number of hits. 

If this will impact a project you maintain or use please do comment on this issue. Many thanks!

**Comment by aseyboldt** [comment]
I regularly use `pycc` as a helper to get profiling information about numba. If I use dynamic jit compiled code I can not see good benchmark information in `perf` or similar tools (see https://github.com/numba/numba/issues/5028), but by compiling a function of interest using pycc and then loading the generated shared library I can get decent profiling info (although debugging symbols are missing, which means I have to practice reading assembly :-) ).
I don't think this is necessarily a good reason not to deprecate it anyway, but I'll miss it if it is gone...

**Comment by stuartarchibald** [comment]
@aseyboldt that's an interesting and unexpected use of `pycc`, thanks for sharing it! This task is just to deprecate `pycc` so as to make it clear to users that it is eventially going to be removed. I hope by the time it is actually removed, Numba will have migrated to using LLVM's ORCJIT which has `perf` support (Numba currently uses MCJIT which does not).

**Comment by ChiCheng45** [comment]
I maintain a scientific software package (currently closed source) which uses the numba AOT compile in pycc so I guess this deprecation will affect us at some point. I understand the reason for the deprecation given the deprecation of distutils and etc. 

But anyway we began using JIT with numba but found that the compile time to be quite long for some of our functions. We also tried the caching functionality but we found this to be quite problematic when running a lot of jobs simultaneously on a HPC with fresh repo so cache files hadn't been generated yet. Numba AOT seemed to be the best thing to do.

We had some ideas about what we should do such as moving to cython (this would be quite time-consuming) or perhaps we'd just fix our conda envs to a specific version of numba. 

But anyway was wondering if you have any suggestions for people who do use pycc a lot in their code.

**Comment by auderson** [comment]
Will we have any substitutions for pycc? AOT functionality is quite essential if we want to use numba in production.

**Comment by stuartarchibald** [comment]
> Will we have any substitutions for pycc? AOT functionality is quite essential if we want to use numba in production.

@auderson yes, as noted in https://github.com/numba/numba/issues/8509#issuecomment-1339629819, a replacement for this functionality will be provided when it is finally removed.

**Comment by stuartarchibald** [comment]
PR #8795 issues pending-deprecation warnings.

**Comment by Hvass-Labs** [comment]
This feature is also useful in cloud computing that don't allow the use of Numba JIT, but you can make a small pre-compiled library of your JIT functions. But I find it difficult to use. Consider the following example:

    from numba.pycc import CC

    cc = CC('test_cc1')

    @cc.export('f', 'f8(f8)')
    def f(x):
        return x + 123.4

    @cc.export('g', 'f8(f8)')
    def g(y):
        return y - 456.7

    @cc.export('h', 'f8(f8, f8)')
    def h(x, y):
        return f(x) * g(y)

    if __name__ == "__main__":
        cc.compile()

When running `python test_cc_src1.py` it gives the following error:

    (...)
    numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
    Untyped global name 'f': Cannot determine Numba type of <class 'function'>

    File "test_cc_src1.py", line 15:
    def h(x, y):
        return f(x) * g(y)
        ^

If I split this into two files so that `f` and `g` go into `test_cc_src1.py` and `h` goes into `test_cc_src2.py`, and then try and compile them in turn, I still get a similar error for the second file. I couldn't figure out how to make this work. Is there a solution?

Perhaps this is something you could improve in a new version of this system.

Thanks!

**Comment by stuartarchibald** [comment]
@Hvass-Labs try decorating e.g. the function `f` with `@njit` as well as `@cc.export(...)` the `CC` compiler should then be able to "see" the function `f`.

**Comment by geofft** [comment]
Is there a place we can follow development of the replacement AOT approach?

We use pycc internally (essentially an HPC use case) to avoid the first-call overhead. In our case I don't think we have any other constraints, so hm, maybe we can work with the existing support for [caching the JIT output](https://numba.readthedocs.io/en/stable/reference/jit-compilation.html#jit-decorator-cache). Do you happen to know if anyone has an example / approach for doing explicit caching during the build, like a plugin for setuptools or poetry or something?

We can deal with the cache being CPU-specific - we'd likely want to set `NUMBA_CPU_NAME` to our baseline CPU anyway for reproducibility, but we can definitely just loop over all the CPUs in our environment.

**Comment by stuartarchibald** [comment]
> Is there a place we can follow development of the replacement AOT approach?

A replacement for #8971 for 2024 will be posted soon and pinned to the top of the issue tracker, it will outline such development.
 
> We use pycc internally (essentially an HPC use case) to avoid the first-call overhead. In our case I don't think we have any other constraints, so hm, maybe we can work with the existing support for [caching the JIT output](https://numba.readthedocs.io/en/stable/reference/jit-compilation.html#jit-decorator-cache). Do you happen to know if anyone has an example / approach for doing explicit caching during the build, like a plugin for setuptools or poetry or something?

I don't know of any such project I'm afraid, it might be worth asking on Numba's discourse forum about approaches others are taking with regards to reusing caches? I'm not sure how reliable such an approach would be in practice.

> We can deal with the cache being CPU-specific - we'd likely want to set `NUMBA_CPU_NAME` to our baseline CPU anyway for reproducibility, but we can definitely just loop over all the CPUs in our environment.

It's also worth nothing that `pycc` is deprecated pending removal, but full deprecation and removal is on the condition that a replacement is available, so it's fine to carry on using it.

**Comment by Ye980226** [comment]
@stuartarchibald 
I hope you're doing well. I'm reaching out to inquire about the latest developments regarding Ahead-Of-Time (AOT) compilation within Numba. As pycc has been deprecated, I am particularly interested in understanding the current alternatives or advancements that have been made in AOT compilation.

Could you please direct me to where I can find the most recent updates or discussions on this topic? Whether it's a specific GitHub issue, PR, or a discussion thread, any guidance would be greatly appreciated.

**Comment by fulcus** [comment]
Hi @stuartarchibald, I have the same question as @Ye980226, any news on this?

**Comment by vavanade** [comment]
Hi! Is there any timeline on the development of the alternatives to `pycc`? I'm assessing the viability of Numba for our use-case (image processing) compared to other approaches like rewriting to C++, Cython etc. Our use-case would benefit from AOT greatly.

**Comment by ca5446001-tech** [comment]
I am using pycc to run pre-compiled numba code from C to eliminate the python overhead and Numba's compile-time memory leaks [6707](https://github.com/numba/numba/issues/6707) from runtime.
